group-coordinator/src/main/java/org/apache/kafka/coordinator/group/share/ShareCoordinatorShard.java

```diff
@@ -267,6 +271,54 @@ public CoordinatorResult<WriteShareGroupStateResponseData, Record> writeState(Re
     return new CoordinatorResult<>(validRecords, responseData);
   }
 
+  /**
+   * This method as called by the ShareCoordinatorService will be provided with
+   * the request data which covers only key i.e. group1:topic1:partition1. The implementation
+   * below was done keeping this in mind.
+   *
+   * @param request - WriteShareGroupStateRequestData for a single key
+   * @param offset  - offset to read from the __share_group_state topic partition
+   * @return CoordinatorResult(records, response)
+   */
+  public ReadShareGroupStateResponseData readState(ReadShareGroupStateRequestData request, Long offset) {
+    log.info("Shard readState request received - {}", request);
+    // records to read (with the key of snapshot type), response to caller
+    // only one key will be there in the request by design
+    Optional<ReadShareGroupStateResponseData> error = maybeGetReadStateError(request);
+    if (error.isPresent()) {
+      return error.get();
+    }
+
+    Uuid topicId = request.topics().get(0).topicId();
+    int partition = request.topics().get(0).partitions().get(0).partition();
+
+    String coordinatorKey = ShareGroupHelper.coordinatorKey(request.groupId(), topicId, partition);
+
+    ShareSnapshotValue snapshotValue = shareStateMap.get(coordinatorKey, offset);
+
+    if (snapshotValue == null) {
+      // Returning an error response as the snapshot value was not found
+      return ReadShareGroupStateResponse.toErrorResponseData(
```

Comment:
We should re-visit this one day and see whether there's a situation in which this might occur where a default success response works. Happy to have it like this in this PR.

Reply:
Sure, I think some discussion is still required for this. Thanks for the approval !